### PR TITLE
Improve error handling

### PR DESCRIPTION
**Pull request by lrettig**
Improves error handling to make downstream error handling easier.
Gives functions better names.

### CODE REVIEW

**Comment by dshulyak** [comment]
would it make sense to implement this checks on consumer side? i thought that this is ok to panic, as consumer of this lib should do this validation or use valid key/path

File: src/lib.rs [hunk 1]
```diff
@@ -19,28 +20,89 @@ pub fn derive_key(
     Box::new(keypair.to_bytes())
 }
 
-/// derive_key_c does the same thing as the above function, but is intended for use over the CFFI.
-/// note that the caller must free() the returned memory as it's not managed/freed here.
+// check for error. if no error, do nothing. if there is an error, print it and return a null ptr.
```
Comment (dshulyak) [inline src/lib.rs:23]: are you sure that this is good option to merge? in smcli this is probably will be seen, but inside smapp it won't be even written as a log. and beside that, will likely fail with some undefined behavior

Comment (dshulyak) [reply src/lib.rs:23]: ah sorry you kept original function as is

**Review by dshulyak — APPROVED** [review]

**Comment by lrettig** [comment]
> would it make sense to implement this checks on consumer side? i thought that this is ok to panic, as consumer of this lib should do this validation or use valid key/path

Good question. The issue with panic here is that AFAICT it's not possible to trap or handle a panic across the FFI boundary - the process just abruptly dies with basically no information about what's wrong. Yes, we could try to recreate the error-checking logic on the consumer side, but that's a duplication of code and logic and if the upstream library changes, the consumer logic will not. I'd be okay with removing the "path too short", "path too long", etc. error checking to the consumer but I'd prefer to leave the other errors (failed to convert string, failed to parse path) in place.

In general the error handling here could be improved a bunch - rather than printing errors, they should be returned, but we'd need a mechanism for returning errors across the FFI boundary. I suppose just returning a string explanation along with the null ptr would be sufficient. I'll open an issue so we can follow up on this later.

**Pull request merged**